### PR TITLE
frontend: created and used GroupedAccountSelector

### PR DESCRIPTION
**Pull request by shonsirsha**
to improve our UX when multiple keystore is connected, we'd like to add a grouping on the account selector.

This account selector is used on the info buy page (`/buy/info`) as well as the receive page (`/accounts/select-receive`) and bitsurance.

Screenshot 2024-01-22 at 11 40 10

;l

d2

d3

### CODE REVIEW

File: frontends/web/src/routes/buy/info.tsx [hunk 1]
```diff
@@ -34,45 +35,50 @@ type TProps = {
...
     }
 
   }, [accounts]);
 
+  useEffect(() => {
+    checkSupportedCoins();
+  }, [checkSupportedCoins]);
+
+
   const maybeProceed = useCallback(async () => {
```
Comment (shonsirsha) [inline frontends/web/src/routes/buy/info.tsx:68]: I wanted to refactor the GroupedAccountSelector to just accept an IAccount object and do all the object / group constructing in that component. 

However, we need the group options outside of that component sometimes like here.

I think it still can be refactored though, I could work on this next. My idea is so that we can just pass an `IAccount[]` and "it just works" (automatically grouped by keystore like we have) without having to do all the grouping in the parent component like currently.

File: frontends/web/src/routes/accounts/select-receive.tsx [hunk 2]
```diff
@@ -16,27 +16,33 @@
...
   useEffect(() => {
-    const options = activeAccounts.map(account => ({ label: account.name, value: account.code, disabled: false, coinCode: account.coinCode } as TOption));
+    const accountNames = getAccountsByKeystore(activeAccounts);
     //setting options without balance
-    setOptions(options);
+    const groupedOpts: TGroupedOption[] = accountNames.map(grouped => ({
+      label: `${grouped.keystore.name} ${isAmbiguiousName(grouped.keystore.name, accountNames) ? `(${grouped.keystore.rootFingerprint})` : ''}`,
+      connected: grouped.keystore.connected,
+      options: grouped.accounts.map((account) => ({ label: account.name, value: account.code, coinCode: account.coinCode, disabled: false })) as TOption[]
+    }));
```
Comment (thisconnect) [inline frontends/web/src/routes/accounts/select-receive.tsx:42]: how about destructing accounts and keystores?

```ts
    const groupedOpts: TGroupedOption[] = accountNames.map(({ accounts, keystore }) => ({
      label: `${keystore.name} ${isAmbiguiousName(keystore.name, accountNames) ? `(${keystore.rootFingerprint})` : ''}`,
      connected: keystore.connected,
      options: accounts.map((account) => ({ label: account.name, value: account.code, coinCode: account.coinCode, disabled: false })) as TOption[]
    }));
```

File: frontends/web/src/routes/bitsurance/account.tsx [hunk 3]
```diff
@@ -48,17 +50,22 @@ export const BitsuranceAccount = ({ code, accounts }: TProps) => {
...
-           || account.bitsuranceStatus === 'canceled'
-           || account.bitsuranceStatus === 'refused'))
-      .map(({ name, code, coinCode }) => (
-        { label: name, value: code, coinCode, disabled: false }
-      ));
-    setBtcAccounts(await setOptionBalances(options));
+
+    // btc accounts that have never been insured, or with a canceled
+    // insurance contract, can be used to make a new contract.
+    const btcAccountz = accounts.filter(account => account.coinCode === 'btc' &&
```
Comment (thisconnect) [inline frontends/web/src/routes/bitsurance/account.tsx:56]: typo btcAccounts :)

File: frontends/web/src/routes/bitsurance/account.tsx [hunk 4]
```diff
@@ -48,17 +50,22 @@ export const BitsuranceAccount = ({ code, accounts }: TProps) => {
...
+        (!account.bitsuranceStatus
+         || account.bitsuranceStatus === 'canceled'
+        || account.bitsuranceStatus === 'refused'));
+
+    const groupedBTCAccounts = getAccountsByKeystore(btcAccountz);
+    const groupedOpts: TGroupedOption[] = groupedBTCAccounts.map(grouped => ({
+      label: `${grouped.keystore.name} ${isAmbiguiousName(grouped.keystore.name, groupedBTCAccounts) ? `(${grouped.keystore.rootFingerprint})` : ''}`,
+      connected: grouped.keystore.connected,
+      options: grouped.accounts.map((account) => ({ label: account.name, value: account.code, coinCode: account.coinCode, disabled: false })) as TOption[]
+    }));
```
Comment (thisconnect) [inline frontends/web/src/routes/bitsurance/account.tsx:66]: is this map callback function the same as in select-receive and buy-info, can you move it to a shared function?

File: frontends/web/src/routes/bitsurance/account.tsx [hunk 5]
```diff
@@ -97,16 +104,17 @@ export const BitsuranceAccount = ({ code, accounts }: TProps) => {
...
                 <div>{t('bitsuranceAccount.noAccount')}</div>
               ) : (
-                <AccountSelector
+                <GroupedAccountSelector
                   title={t('bitsuranceAccount.select')}
-                  options={btcAccounts}
                   disabled={disabled}
+                  options={groupedOptions}
                   selected={selected}
                   onChange={handleChangeAccount}
```
Comment (thisconnect) [inline frontends/web/src/routes/bitsurance/account.tsx:115]: here you didn't rewrite onChange, but in the 2 other similar components you have  `onChange={(selected) => setSelected(selected)}` and `onChange={setCode}`.

**Review by thisconnect — CHANGES_REQUESTED** [review]

File: frontends/web/src/components/groupedaccountselector/groupedaccountselector.tsx [hunk 6]
```diff
@@ -65,7 +66,7 @@ const SelectSingleValue: FunctionComponent<SingleValueProps<TOption>> = (props)
   );
 };
 
-const SelectOption: FunctionComponent<OptionProps<TOption>> = (props) => {
+const SelectOption = (props: OptionProps<TOption, false>) => {
```
Comment (thisconnect) [inline frontends/web/src/components/groupedaccountselector/groupedaccountselector.tsx:69]: why is this needed?

**Comment by thisconnect** [comment]
When going to `/buy/info` In testnet (`make servewallet`) the keystore prompts directly. 

It works normally in  mainnet ( with `make servewallet-mainnet`).

it doesn't happen for me in master.

File: frontends/web/src/routes/bitsurance/account.tsx (hunk 5 above)
Comment (shonsirsha) [reply frontends/web/src/routes/bitsurance/account.tsx:115]: oups, good point. this one comes from the Bitsurance staging merge,  I'll just change this to setState. Same with the () => setSelected, will change to setSelected. Ty!

File: frontends/web/src/components/groupedaccountselector/groupedaccountselector.tsx (hunk 6 above)
Comment (shonsirsha) [reply frontends/web/src/components/groupedaccountselector/groupedaccountselector.tsx:69]: I only wanted to remove the `FunctionComponent` as it's not needed (easier to read without it, right?). I'll omit the `false` as thats the default value anyway (for `isMulti` - for multi select in react-select).

File: frontends/web/src/routes/bitsurance/account.tsx (hunk 4 above)
Comment (shonsirsha) [reply frontends/web/src/routes/bitsurance/account.tsx:66]: Yeah, it's related to [this](https://github.com/digitalbitbox/bitbox-wallet-app/pull/2497#discussion_r1461681393). I'll refactor.

File: frontends/web/src/routes/accounts/select-receive.tsx (hunk 2 above)
Comment (shonsirsha) [reply frontends/web/src/routes/accounts/select-receive.tsx:42]: tbh I prefer it this way. To me it's clearer which `name` it is for example if we explicitly do `keystore.name` than just `name` knowing in this block we also have `account.name`. Also clearer for me personally to see that `rootFingerPrint` for ex, comes from keystore by doing`keystore.`. 

 But if you strongly prefer destructured, we could do that too ofc :)

**Comment by shonsirsha** [comment]
> When going to `/buy/info` In testnet (`make servewallet`) the keystore prompts directly.
> 
> It works normally in mainnet ( with `make servewallet-mainnet`).
> 
> it doesn't happen for me in master.

I don't think I can reproduce this. @Beerosagos, would you be able to check?

**Comment by shonsirsha** [comment]
@thisconnect thanks for the review. I refactored to achieve [this](https://github.com/digitalbitbox/bitbox-wallet-app/pull/2497#discussion_r1461681393), which should account for your [comment](https://github.com/digitalbitbox/bitbox-wallet-app/pull/2497#discussion_r1462232630). I think it looks a lot cleaner now, and easier to use as we can just plug-in an `IAccount[]` into the `GroupedAccountSelector`. PTAL :)

**Comment by thisconnect** [comment]
> When going to /buy/info In testnet (make servewallet) the keystore prompts directly.

thanks @shonsirsha and @Beerosagos for looking into this. looks like it is correct behavior if there is only 1 account that is supported by the exchanges (i.e. in testnet only moonpay sells TBTC).

but weird that i had it as I always had 3 watch-only wallets with at least TBTC 1 account each. maybe it was loading weirdly?

anyway I can't reproduce the issue anymore since this branch was rebased. soo all good 👍

**Review by thisconnect — COMMENTED** [review]
LGTM with small nits

I'll do some more testing.

**Review by thisconnect — APPROVED** [review]
LGTM
